\documentclass{bioinfo}

\usepackage{relsize} % for Cpp
\usepackage{xspace} % for Cpp

\newcommand{\wfmash}{wfmash}
\newcommand{\Rplus}{\protect\hspace{-.1em}\protect\raisebox{.35ex}{\smaller{\smaller\textbf{+}}}}
\newcommand{\pp}{\Rplus\Rplus}
\newcommand{\Cpp}{\mbox{C\Rplus\Rplus}\xspace}

\copyrightyear{2021} \pubyear{2021}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Original Paper}

\begin{document}
    \firstpage{1}

    \subtitle{Sequence analysis}

%\title[short Title]{Wavefront inception scales global pairwise alignment to whole eukaryotic chromosomes}
%\title[short Title]{Wavefront inception enables global pairwise alignment to whole eukaryotic chromosomes}
    \title[wfmash: whole-cromosome pairwise alignment]{Whole-chromosome pairwise alignment using a hierarchical wavefront algorithm}
    \author[Sample \textit{et~al}.]{
        Corresponding Author\,$^{\text{\sfb 1,}*}$,
        Co-Author\,$^{\text{\sfb 2}}$ and
        Co-Author\,$^{\text{\sfb 3,}*}$
    }
    \address{
        $^{\text{\sf 1}}$Department, Institution, City, Post Code, Country \\
        $^{\text{\sf 2}}$Department, Institution, City, Post Code, Country \\
        $^{\text{\sf 3}}$Department, Institution, City, Post Code, Country
    }

    \corresp{$^\ast$To whom correspondence should be addressed.}

    \history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

    \editor{Associate Editor: XXXXXXX}

    \abstract{
        \textbf{Motivation:} Pairwise alignment of sequences is an important step in many bioinformatics analyses,
        including pangenomes building. Pangenomes are sequence models able to provide a full representation of the
        mutual alignment of collections of genomes. The time and memory required to compute a pairwise alignment
        increase quadratically with the sequence length, making it impractical to directly align very long sequences
        without applying heuristic approaches to determine possible syntenic regions in the sequences. Nevertheless,
        with the advances in sequencing technologies, new genome assemblies are produced at a high rate, pressing for
        the development of tools able to align sequences of the order of tens of megabases long.
        \\
        \textbf{Results:} Here we present \wfmash\, a new gap-affine pairwise aligner designed to align DNA sequences
        at a whole-chromosome scale. \wfmash\ applies a hierarchical implementation of the wavefront alignment
        algorithm to guide the alignment of very long sequences. It scales efficiently to large eukaryotic
        chromosomes, allowing users to perform pairwise alignment of thousands of large genomes using little
        time and memory.
        \\
        \textbf{Availability:} \wfmash\ is written in \Cpp\ programming language. Source code and user manual are freely available at
        \href{https://github.com/ekg/wfmash}{https://github.com/ekg/wfmash}.
        \\
        \textbf{Contact:}
        \href{andreaguarracino@outlook.com}{andreaguarracino@outlook.com},
        \href{erik.garrison@gmail.com}{erik.garrison@gmail.com}
        \\
        \textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics} online.
    }

    \maketitle


    \section{Introduction}

    Many biological analyses take advantage of aligning DNA sequences, ranging from read~mapping
    \citep{22388286, BWA_MEM, 23103880} to variant detection~\citep{21478889}, as well as de~novo
    assembly \citep{19251739} and pangenome building~\citep{33177663, 33066802}. Pairwise sequence alignment
    can be applied to identify similar regions that may indicate functional, structural, and/or evolutionary
    relationships between two biological sequences. In this context, pangenomes model the full set
    of genomic elements in a given species or clade~\citep{32453966}. These data structures encode the mutual
    relationships between all the genomes represented, in contrast to reference-based approaches which
    relate sequences to a particular genome, chosen as reference. The unbiased approach allows studying
    the entire genetic diversity of a population. This opportunity, in combination with the massive amount
    of data available, made possible thanks to the advancement in sequencing technology, presses for the
    development of tools able to compute pairwise alignment of very long sequences.

    To face the upcoming challenges in pangenomes building, we have developed \wfmash\, a new gap-affine
    pairwise tool for aligning DNA sequences at a whole-chromosome scale. It supports split-read alignment
    and gap-affine penalties for insertion and deletions. \wfmash\ scales efficiently to large eukaryotic
    chromosomes while requiring little time and memory, making it suitable to be applied on the pairwise
    alignment of hundreds of genomes. \wfmash\ is a DNA sequence aligner based on the integration of
    MashMap~\citep{30423094}, a fast approximate aligner for computing local alignment boundaries between
    long DNA sequences, and the wavefront alignment algorithm (WFA)~\citep{32915952}, an exact gap-affine algorithm
    that takes advantage of homologous regions between the sequences to accelerate the alignment process.

    %TODO Some sentences of why it exists with respect to Minimap2/Winnomap2, explaining the limitations of those tools in the pairwise alignment task.


%\enlargethispage{12pt}


    \section{Methods}

    \wfmash\ first applies a locality-sensitive hashing, from MashMap, to rapidly determine syntenic region
    boundaries between long DNA sequences. Then, a hierarchical implementation of the WFA allows computing the
    base-level global alignment of the identified mappings.

    \subsection{Approximate mapping}

    Each query sequence is broken into non-overlapping pieces of the requested length. These segments are then
    mapped using MashMap's sliding MinHash mapping algorithm~\citep{30423094}. We extended MashMap, fixing bugs
    and incorporating the robust winnowing \citep{Schleimer S. et al.  (2003)} in the minimizer downsampling.
    Robust sampling avoids taking too many minimizers in low-complexity substrings, yielding improvements in
    runtime and memory-usage without affecting accuracy~\citep{32657365}.

    %TODO THINGS ON THE FILTERING?
    %TODO THING ON THE SPLIT READ AND/OR THE MERGING?

    In addition to the segment length, the minimum segment estimated identity and the maximum number of
    mappings to report for each segment can be specified in the mapping process. These settings allow users
    to precisely define the mapping space to consider, specifying the characteristics of homologies to compute.
    Each mapping location is then used as a target for base-level alignment using a hierarchical implementation
    of the WFA.

    \subsection{Base-level alignment}

    The time and memory required to compute the base-level alignment increases quadratically with the sequence
    length. The WFA provides an efficient way to decrease the amount of computation required to obtain the
    optimal alignment between two sequences, reducing the cost to be quadratic in the alignment penalty score
    of the optimal alignment~\citep{32915952}. This means that the algorithm is very efficient in aligning
    similar homologous sequences (i.e., sequences with a low alignment penalty score), but high divergence
    genomes and/or noisy long reads can increase its memory and runtime costs.

    To avoid such limitation, wfmash applies a hierarchical WFA, exploiting the WFA to guide the alignment
    process, but keeping the largest alignment problem size small. Rather than directly aligning the whole
    sequences, it aligns them to each other in small pieces W-bps long by applying a global WFA alignment
    (Figure~1\vphantom{\ref{fig:1}}). The whole global alignment is then computed over the full
    dynamic-programming matrix (high order DP-matrix) at \textit{W}-bps resolution.
    Each cell in the high order DP-matrix corresponds to the alignment of a specific pair of fragments
    \textit{W}-bps long from the two sequences to align. To determine if each cell is a match, and then guiding the
    \textit{W}-bps resolution alignment, the global alignment using the standard WFA is performed between the
    two fragments.


    \begin{figure}[!tpb]%figure1
        \fboxsep=0pt\colorbox{gray}{
            \begin{minipage}[t]{235pt}
                \vbox to 100pt{\vfill\hbox to
                235pt{\hfill\fontsize{24pt}{24pt}\selectfont FPO\hfill}\vfill}
            \end{minipage}}
            %\centerline{\includegraphics{fig01.eps}}
        \caption{Generate a cool plot using R as a base to create a cool image.}\label{fig:1}
    \end{figure}

%\begin{figure}[!tpb]%figure2
%%\centerline{\includegraphics{fig02.eps}}
%\caption{Caption, caption.}\label{fig:02}
%\end{figure}

    Finally, the \textit{W}-bps resolution traceback is applied to determine the set of base-level alignments
    on the optimal path in the high order DP-matrix. During the process, a further step is performed for
    refining the alignment. Indeed, inaccurate mapping estimation can lead to under-alignment at the beginning
    and the end of the aligned syntenic region. Therefore, edlib is applied by computing semi-global alignments
    to resolve the mapping boundaries.
    The hierarchical implementation requires only the memory to align a syntenic region at \textit{W}-bps resolution,
    limiting the runtime and the memory of the standard WFA by applying it to fragments \textit{W}-bps long.
    This approach is more flexible than using a fixed-width band (it effectively has a variable bandwidth), requires
    no heuristic seeding step, and can benefit from parallel exploration of the wavefront.
    %TODO 'this approach is more flexible than using a fixed-width band: Not sure about writing it. I would remove it, else we should demonstrate it.
    %TODO 'and can benefit from parallel exploration of the wavefront.' Can we?



    \section{Results}

    Text Text Text Text Text Text Text Text Text Text Text Text Text
    Text Text Text Text Text Text Text Text.


    \section{Discussion}

    We implemented a novel gap-affine pairwise aligner, \wfmash\, to accelerate the computation of the mutual
    alignment of collections of genomes, a step required for building pangenome models. We have demonstrated that
    it efficiently performs with contigs representing full human chromosomes of 88 phased haplotypes from the
    Human Pangenome Reference Consortium year 1 assembly. Indeed, thanks to its efficiency, wfmash is already
    successfully applied in our pangenome graphs building pipeline (\citep{pggb}). No less important, pairwise
    alignment is a central step of many bioinformatics applications, making our aligner a scalable solution to face the
    increasing yields of sequencing technologies in the coming years.

    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%     please remove the " % " symbol from \centerline{\includegraphics{fig01.eps}}
%     as it may ignore the figures.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}


    \begin{thebibliography}{}

        \bibitem[Bofelli {\it et~al}., 2000]{Boffelli03}
        Bofelli,F., Name2, Name3 (2003) Article title, {\it Journal Name}, {\bf 199}, 133-154.

        \bibitem[Bag {\it et~al}., 2001]{Bag01}
        Bag,M., Name2, Name3 (2001) Article title, {\it Journal Name}, {\bf 99}, 33-54.

        \bibitem[Yoo \textit{et~al}., 2003]{Yoo03}
        Yoo,M.S. \textit{et~al}. (2003) Oxidative stress regulated genes
        in nigral dopaminergic neurnol cell: correlation with the known
        pathology in Parkinson's disease. \textit{Brain Res. Mol. Brain
        Res.}, \textbf{110}(Suppl. 1), 76--84.

        \bibitem[Lehmann, 1986]{Leh86}
        Lehmann,E.L. (1986) Chapter title. \textit{Book Title}. Vol.~1, 2nd edn. Springer-Verlag, New York.

        \bibitem[Crenshaw and Jones, 2003]{Cre03}
        Crenshaw, B.,III, and Jones, W.B.,Jr (2003) The future of clinical
        cancer management: one tumor, one chip. \textit{Bioinformatics},
        doi:10.1093/bioinformatics/btn000.

        \bibitem[Auhtor \textit{et~al}. (2000)]{Aut00}
        Auhtor,A.B. \textit{et~al}. (2000) Chapter title. In Smith, A.C.
        (ed.), \textit{Book Title}, 2nd edn. Publisher, Location, Vol. 1, pp.
        ???--???.

        \bibitem[Bardet, 1920]{Bar20}
        Bardet, G. (1920) Sur un syndrome d'obesite infantile avec
        polydactylie et retinite pigmentaire (contribution a l'etude des
        formes cliniques de l'obesite hypophysaire). PhD Thesis, name of
        institution, Paris, France.

    \end{thebibliography}
\end{document}
