# All-vs-All

```shell
DIR_BASE=/lizardfs/guarracino/wfmash-paper

FASTIX=/lizardfs/guarracino/tools/fastix/target/release/fastix-331c1159ea16625ee79d1a82522e800c99206834
WFMASH=/home/guarracino/tools/wfmash/build/bin/wfmash-e7850f9e9ba743a350b36b5e51ae8c414fba717d

conda create --prefix /lizardfs/guarracino/condatools/compleasm/0.2.5/ -c conda-forge -c bioconda compleasm=0.2.5 -y
```

## Data

```shell
mkdir -p $DIR_BASE/assemblies/scerevisiae
cd $DIR_BASE/assemblies/scerevisiae

wget -c https://raw.githubusercontent.com/pangenome/pggb/master/docs/data/scerevisiae.yprp.urls
cat scerevisiae.yprp.urls | grep mt -v | parallel -j 4 'wget -q {} && echo got {}'

## PanSN
ls *.fa.gz | while read FASTA; do
    SAMPLE=$(echo $FASTA | cut -f 1 -d '.');
    echo $SAMPLE

    zcat $FASTA | sed "s/^>/>$SAMPLE#1#/g" | bgzip -@ 48 -l 9 > $SAMPLE.fasta.gz && samtools faidx $SAMPLE.fasta.gz
done
rm *.fa.gz

zcat *.fasta.gz | bgzip -@ 48 -l 9 > scerevisiae7.fa.gz && samtools faidx scerevisiae7.fa.gz
```

## BUSCO genes

```shell
mkdir -p $DIR_BASE/busco
cd $DIR_BASE/busco

conda activate /lizardfs/guarracino/condatools/compleasm/0.2.5 

compleasm download -L $DIR_BASE/busco/databases saccharomycetes

mkdir -p $DIR_BASE/busco/scerevisiae
cd $DIR_BASE/busco/scerevisiae
ls $DIR_BASE/assemblies/scerevisiae/*fasta.gz | while read FASTA; do
    SAMPLE=$(basename $FASTA .fasta.gz)
    echo $SAMPLE
    
    DIR_OUTPUT=$DIR_BASE/busco/scerevisiae/$SAMPLE
    mkdir -p $DIR_OUTPUT
    sbatch -c 48 -p workers  --job-name busco-saccharomycetes-$SAMPLE --wrap "hostname; cd /scratch; compleasm run -a $FASTA -o $DIR_OUTPUT -t 48 -l saccharomycetes -L $DIR_BASE/busco/databases; mv $DIR_OUTPUT/summary.txt $DIR_OUTPUT/$SAMPLE.summary.saccharomycetes.txt"
done

conda deactivate

# Table to BED
ls $DIR_BASE/assemblies/scerevisiae/*fasta.gz | while read FASTA; do
    SAMPLE=$(basename $FASTA .fasta.gz)
    echo $SAMPLE
    
    DIR_OUTPUT=$DIR_BASE/busco/scerevisiae/$SAMPLE
    grep Missing -v $DIR_OUTPUT/saccharomycetes_odb10/full_table.tsv | sed '1d' | awk -v OFS='\t' '{print($3,$4,$5,$1,$7,$6,$2)}' > $DIR_OUTPUT/saccharomycetes_odb10/$SAMPLE.busco-genes.bed
done
```

## Alignment

```shell
mkdir -p $DIR_BASE/alignment/scerevisiae
cd $DIR_BASE/alignment/scerevisiae

PAF=$DIR_BASE/alignment/scerevisiae/DBVPG6044-vs-S288C.paf
$WFMASH $DIR_BASE/assemblies/scerevisiae/S288C.fasta.gz $DIR_BASE/assemblies/scerevisiae/DBVPG6044.fasta.gz -t 48 > $PAF

PATH_QUERY_BED=$DIR_BASE/busco/scerevisiae/DBVPG6044/saccharomycetes_odb10/DBVPG6044.busco-genes.bed
PATH_TARGET_BED=$DIR_BASE/busco/scerevisiae/S288C/saccharomycetes_odb10/S288C.busco-genes.bed



# Number of GENCODE genes in the target
total_genes_in_target=$(cat $path_gencode_genes_target | wc -l)

#cat alignment/$prefix.paf | awk -v OFS='\t' '{print $1, $3, $4, "", "", $5}' >alignment/$prefix.$query.bed
cat alignment/$prefix.paf | awk -v OFS='\t' '{print $6, $8, $9, "", "", "+"}' | sed 's/grch38#//g' >alignment/$prefix.$target.bed

# Number of GENCODE genes not entirely covered
missing_genes_in_target=$(bedtools subtract -a $path_gencode_genes_target -b alignment/$prefix.$target.bed | cut -f 4 | sort | uniq | wc -l)

missing_gene_ratio_target=$(echo "scale=4; 1 - $missing_genes_in_target / $total_genes_in_target" | bc)

echo $query $target $total_genes_in_target $s $l $p $n $w $missing_gene_ratio_target $missing_genes_in_target | tr ' ' '\t' >>gencode_evaluation.mapping.tsv
```

## Evaluation

### BUSCO coverage

### BUSCO matches
