# All-vs-All

```shell
DIR_BASE=/lizardfs/guarracino/wfmash-paper

FASTIX=/lizardfs/guarracino/tools/fastix/target/release/fastix-331c1159ea16625ee79d1a82522e800c99206834
WFMASH=/home/guarracino/tools/wfmash/build/bin/wfmash-e7850f9e9ba743a350b36b5e51ae8c414fba717d

conda create --prefix /lizardfs/guarracino/condatools/compleasm/0.2.5/ -c conda-forge -c bioconda compleasm=0.2.5 -y
```

## Data

```shell
mkdir -p $DIR_BASE/assemblies/scerevisiae
cd $DIR_BASE/assemblies/scerevisiae

wget -c https://raw.githubusercontent.com/pangenome/pggb/master/docs/data/scerevisiae.yprp.urls
cat scerevisiae.yprp.urls | grep mt -v | parallel -j 4 'wget -q {} && echo got {}'

# PanSN
ls *.fa.gz | while read FASTA; do
    SAMPLE=$(echo $FASTA | cut -f 1 -d '.');
    echo $SAMPLE

    zcat $FASTA | sed "s/^>/>$SAMPLE#1#/g" | bgzip -@ 48 -l 9 > $SAMPLE.fasta.gz && samtools faidx $SAMPLE.fasta.gz
done
rm *.fa.gz

zcat *.fasta.gz | bgzip -@ 48 -l 9 > scerevisiae7.fa.gz && samtools faidx scerevisiae7.fa.gz
```

## BUSCO genes

```shell
mkdir -p $DIR_BASE/busco
cd $DIR_BASE/busco

conda activate /lizardfs/guarracino/condatools/compleasm/0.2.5 

compleasm download -L $DIR_BASE/busco/databases saccharomycetes

mkdir -p $DIR_BASE/busco/scerevisiae
cd $DIR_BASE/busco/scerevisiae
ls $DIR_BASE/assemblies/scerevisiae/*fasta.gz | while read FASTA; do
    SAMPLE=$(basename $FASTA .fasta.gz)
    echo $SAMPLE
    
    DIR_OUTPUT=$DIR_BASE/busco/scerevisiae/$SAMPLE
    mkdir -p $DIR_OUTPUT
    sbatch -c 48 -p workers  --job-name busco-saccharomycetes-$SAMPLE --wrap "hostname; cd /scratch; compleasm run -a $FASTA -o $DIR_OUTPUT -t 48 -l saccharomycetes -L $DIR_BASE/busco/databases; mv $DIR_OUTPUT/summary.txt $DIR_OUTPUT/$SAMPLE.summary.saccharomycetes.txt"
done

conda deactivate

# Table to BED
ls $DIR_BASE/assemblies/scerevisiae/*fasta.gz | while read FASTA; do
    SAMPLE=$(basename $FASTA .fasta.gz)
    echo $SAMPLE
    
    DIR_OUTPUT=$DIR_BASE/busco/scerevisiae/$SAMPLE
    grep Missing -v $DIR_OUTPUT/saccharomycetes_odb10/full_table.tsv | sed '1d' | awk -v OFS='\t' '{print($3,$4,$5,$1,$7,$6,$2)}' > $DIR_OUTPUT/saccharomycetes_odb10/$SAMPLE.busco-genes.bed
done
```

## Alignment

```shell
mkdir -p $DIR_BASE/alignment/scerevisiae
cd $DIR_BASE/alignment/scerevisiae

#### FOR DEVELOPMENT############################################################
samtools faidx $DIR_BASE/assemblies/scerevisiae/S288C.fasta.gz $(head -n 1 $DIR_BASE/assemblies/scerevisiae/S288C.fasta.gz.fai | cut -f 1) | bgzip -@ 48 -l 9 > S288C_1_chrI.fa.gz && samtools faidx S288C_1_chrI.fa.gz
#samtools faidx $DIR_BASE/assemblies/scerevisiae/DBVPG6044.fasta.gz $(head -n 1 $DIR_BASE/assemblies/scerevisiae/DBVPG6044.fasta.gz.fai | cut -f 1) -i | bgzip -@ 48 -l 9 > DBVPG6044_1_chrI.fa.gz && samtools faidx DBVPG6044_1_chrI.fa.gz
read CHR LENGTH <<< $(head -n 1 $DIR_BASE/assemblies/scerevisiae/DBVPG6044.fasta.gz.fai | cut -f 1,2)
# Calculate the midpoint of the chromosome (assuming integer division truncates the fraction)
MIDPOINT=$((LENGTH / 2))
# Extract and reverse complement the first half of the chromosome
cat <(samtools faidx $DIR_BASE/assemblies/scerevisiae/DBVPG6044.fasta.gz $CHR:1-$MIDPOINT -i) <(samtools faidx $DIR_BASE/assemblies/scerevisiae/DBVPG6044.fasta.gz $CHR:$((MIDPOINT + 1))-$LENGTH) | bgzip -@ 48 -l 9 > DBVPG6044_1_chrI.fa.gz && samtools faidx DBVPG6044_1_chrI.fa.gz

PATH_QUERY_BED=$DIR_BASE/busco/scerevisiae/DBVPG6044/saccharomycetes_odb10/DBVPG6044.busco-genes.bed
PATH_TARGET_BED=$DIR_BASE/busco/scerevisiae/S288C/saccharomycetes_odb10/S288C.busco-genes.bed

PATH_PAF=$DIR_BASE/alignment/scerevisiae/DBVPG6044-vs-S288C.paf
$WFMASH $DIR_BASE/assemblies/scerevisiae/S288C.fasta.gz $DIR_BASE/assemblies/scerevisiae/DBVPG6044.fasta.gz -t 48 > $PATH_PAF

#PATH_PAF=$DIR_BASE/alignment/scerevisiae/DBVPG6044-vs-S288C.chrI.paf
#$WFMASH S288C_1_chrI.fa.gz DBVPG6044_1_chrI.fa.gz -t 48 -s 100 -p 50 > $PATH_PAF
#PATH_QUERY_BED=/lizardfs/guarracino/wfmash-paper/alignment/scerevisiae/busco_interted_chr/saccharomycetes_odb10/DBVPG6044_1_chrI.rev.bed
################################################################################

PATH_PAF=$DIR_BASE/alignment/scerevisiae/scerevisiae7.paf
$WFMASH $DIR_BASE/assemblies/scerevisiae/scerevisiae7.fa.gz -t 48 --lower-triangular > $PATH_PAF

PATH_BUSCO_BED=$DIR_BASE/alignment/scerevisiae/scerevisiae.busco-genes.single.bed
cat $DIR_BASE/busco/scerevisiae/*/saccharomycetes_odb10/*.busco-genes.bed | grep 'Single$' > $PATH_BUSCO_BED # Put all the BUSCO genes in a single file

bash $DIR_BASE/scripts/check_features.sh $PATH_PAF $PATH_BUSCO_BED 59 ./scerevisiae7.m50 /scratch

############################################################################################################################################
# Working implementation
########################
# NAME=$(basename $PAF .paf)
# bedtools intersect -a <(awk -v OFS='\t' '{print($1,$3,$4,$0)}' $PAF) -b <(grep 'Single$' $PATH_QUERY_BED)  -wa -wb | cut -f 4- > $NAME.query.paf
# bedtools intersect -a <(awk -v OFS='\t' '{print($6,$8,$9,$0)}' $PAF) -b <(grep 'Single$' $PATH_TARGET_BED) -wa -wb | cut -f 4- > $NAME.target.paf

# # Join the two files
# join -1 1 -2 1 \
#     <(awk -v OFS='\t' '{concat=$1; for(i=2;i<=20;i++) concat=concat "___" $i; concat=concat "___" $24; print(concat,$21,$22,$23,$25,$26,$27)}' $NAME.query.paf  | sort -T /scratch) \
#     <(awk -v OFS='\t' '{concat=$1; for(i=2;i<=20;i++) concat=concat "___" $i; concat=concat "___" $24; print(concat,$21,$22,$23,$25,$26,$27)}' $NAME.target.paf | sort -T /scratch) |
#     # Deconcatenate the joined column back into original columns
#     # What happens if the gene is covered by more mappings in one genome than in the other???????????????????????????????????
#     awk '{
#         # Split the first field into array `a` using the "|" delimiter
#         n=split($1, a, "___");
#         # Reconstruct the original columns
#         for(i=1; i<=20; i++) printf("%s\t", a[i]);
#         printf("%s\t%s\t%s\t%s\t%s\t%s\t%s\t", $2, $3, $4, a[21], $5, $6, $7);
#         printf("%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $8, $9, $10, a[21], $11, $12, $13);
#     }' | cut -f 1-12,20- | pigz -9 > $NAME.joined.paf.gz

# # Number of BUSCO genes not covered by the mapping on both genomes
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

# zcat $NAME.joined.paf.gz | python3 x.py > alignments-checked.tsv
############################################################################################################################################


# Check with SafFire
GENE=2456at4891
zgrep $GENE DBVPG6044-vs-S288C.chrI.joined.paf.gz | cut -f 1-13 > $GENE.paf
zgrep $GENE DBVPG6044-vs-S288C.chrI.joined.paf.gz | rev | cut -f 8-14 | rev > $GENE.in-query.bed
zgrep $GENE DBVPG6044-vs-S288C.chrI.joined.paf.gz | rev | cut -f 1-7  | rev > $GENE.in-target.bed
~/git/rustybam/target/release/rb stats --paf $GENE.paf > $GENE.saffire
~/git/rustybam/target/release/rb break-paf --max-size 1 $GENE.paf | ~/git/rustybam/target/release/rb stats --paf > $GENE.saffire
~/git/rustybam/target/release/rb liftover <(zgrep $GENE DBVPG6044-vs-S288C.chrI.joined.paf.gz | cut -f 1-13) --bed <(zgrep $GENE DBVPG6044-vs-S288C.chrI.joined.paf.gz | rev | cut -f 1-7 | rev)
~/git/rustybam/target/release/rb liftover <(zgrep $GENE DBVPG6044-vs-S288C.chrI.joined.paf.gz | cut -f 1-13) --bed <(zgrep $GENE DBVPG6044-vs-S288C.chrI.joined.paf.gz | rev | cut -f 8-14| rev) --qbed

############################################################################################################################################
# Slower, but more controllable implementation
##############################################
# # Common single BUSCO genes
# (grep 'Single$' $PATH_QUERY_BED | cut -f 4 | sort; grep 'Single$' $PATH_TARGET_BED | cut -f 4 | sort) | sort | uniq -d > common.busco-genes.txt

# # For each gene, get the mappings that cover them
# # Ignore the strand because it depends on the gene too. For example, 0at4891 is on - for both genomes, and query and target are aligned with the same orientation, so no problem
# cat common.busco-genes.txt | while read GENE; do
#     #echo $GENE >&2

#     # CAN WE GENERATE THE INTERSECTION FOR ALL GENES AT ONCE?????????????????????????????????????????????????????????????????????????????????
#     grep $GENE $PATH_QUERY_BED -w | bedtools intersect -a <(awk -v OFS='\t' '{print($1,$3,$4,$0)}' $PAF) -b - -wa -wb | cut -f 4- | sort > $GENE.query.paf
#     grep $GENE $PATH_TARGET_BED -w | bedtools intersect -a <(awk -v OFS='\t' '{print($6,$8,$9,$0)}' $PAF) -b - -wa -wb | cut -f 4- | sort > $GENE.target.paf

#     # should_continue=false
#     # if [ ! -s "$GENE.query.paf" ]; then
#     #     echo "Warning: $GENE.query.paf is empty" >&2
#     #     should_continue=true
#     # fi
#     # if [ ! -s "$GENE.target.paf" ]; then
#     #     echo "Warning: $GENE.target.paf is empty" >&2
#     #     should_continue=true
#     # fi
#     # if [ "$should_continue" = true ]; then
#     #     continue
#     # fi
    
#     # Check if either file has more than one row and emit a warning if so
#     if [ $(wc -l < "$GENE.query.paf") -gt 1 ] || [ $(wc -l < "$GENE.target.paf") -gt 1 ]; then
#         echo "Warning: More than one mapping found for $GENE, skipping..." >&2
#         continue # Skip to the next GENE
#     fi

#     # What happens if the gene is covered by more mappings in one genome than in the other?
#     python -c "gene=\"$GENE\"; f1=open(f'{gene}.query.paf'); f2=open(f'{gene}.target.paf'); d1={}; d2={}; [d1.setdefault(' '.join(line.split()[:-7]), []).append(line.split()[-7:]) for line in f1]; [d2.setdefault(' '.join(line.split()[:-7]), []).append(line.split()[-7:]) for line in f2]; [print(k, *v[0], *d2[k][0]) for k, v in d1.items() if k in d2]; f1.close(); f2.close()" | cut -f 1-12,20- -d ' ' | tr ' ' '\t'
    
#     rm  $GENE.query.paf $GENE.target.paf
# done | pigz -9 > alignments-to-check.paf.gz

# # Number of BUSCO genes not covered by the mapping on both genomes
# comm -23 common.busco-genes.txt <(zcat alignments-to-check.paf.gz | rev | cut -f 4 | rev) | wc -l

# zcat alignments-to-check.paf.gz | python3 x.py > alignments-checked.tsv
############################################################################################################################################



~/git/rustybam/target/release/rb break-paf --max-size 1 $GENE.paf | ~/git/rustybam/target/release/rb stats --paf > $GENE.saffire

```



# Number of GENCODE genes in the target
total_genes_in_target=$(cat $path_gencode_genes_target | wc -l)

#cat alignment/$prefix.paf | awk -v OFS='\t' '{print $1, $3, $4, "", "", $5}' >alignment/$prefix.$query.bed
cat alignment/$prefix.paf | awk -v OFS='\t' '{print $6, $8, $9, "", "", "+"}' | sed 's/grch38#//g' >alignment/$prefix.$target.bed

# Number of GENCODE genes not entirely covered
missing_genes_in_target=$(bedtools subtract -a $path_gencode_genes_target -b alignment/$prefix.$target.bed | cut -f 4 | sort | uniq | wc -l)

missing_gene_ratio_target=$(echo "scale=4; 1 - $missing_genes_in_target / $total_genes_in_target" | bc)

echo $query $target $total_genes_in_target $s $l $p $n $w $missing_gene_ratio_target $missing_genes_in_target | tr ' ' '\t' >>gencode_evaluation.mapping.tsv
```

## Evaluation

### BUSCO coverage

### BUSCO matches
